trunk/wadtex/wadtex.6
.TH WADTEX 6 "2000-10-26"

.SH NAME
wadtex \- extract textures from WAD files into ppm graphic files

.SH SYNOPSIS
.BR wadtex
.RB [ "\-c \fIpalette\fR" ]
.RB [ "\-i \fIiwadfile\fR" ]
.RB [ \-npt ]
.RB [ \-m | \-M ]
.RB [ \-w\ \fIexpr\fP ]
.RI [ wadfile \ ...]

.SH DISCLAIMER
This program is furnished on an AS IS basis for the fun of it.
id Software, Raven Software and Activision do not support this program.

.SH DESCRIPTION
This program loads one or more wad files (either IWAD or PWAD) and extracts
some or all of the textures they contain to individual files, in packed PPM
(\fBP6\fP) format.  The files are put in the \fB./textures/\fP directory. Thus,
a texture named \fBFRED\fP would be saved as \fB./textures/fred.ppm\fP. If
the \fB./textures\fP directory does not already exist, it is created.

wadtex tries to emulate as closely as possible the behaviour of Doom with
respect to the handling of wads. The list of textures is obtained from
the \fBTEXTURE1\fP lump (and \fBTEXTURE2\fP if it exists). The patch numbers
in \fBTEXTURE1\fP and \fBTEXTURE2\fP are translated to patch names according to
the contents of the \fBPNAMES\fP lump. The patches are located by scanning the
wad directories in a brute force manner. In particular, no attention is paid to
labels like \fB?_START\fP, \fB?_END\fP, \fBMAP??\fP, etc.

The lumps found in the pwads override the lumps in the iwad. If the same lump
is present in more than one pwad, the pwad that comes last on the command line
takes precedence.

If a texture cannot be extracted because it is too large or there is not enough
memory or too many of its patches are missing, it's just skipped with a warning
and the exit code will be non-zero. Most other errors, including write errors,
cause wadtex to abort immediately.

A zero return code guarantees that all textures were extracted but not that
they were all complete ; there could have been missing patches. If you want to
use the exit code as proof of completeness, specify \fB\-m\fP.

.SH OPTIONS
.TP
.BI \-c \ palette
By default the program uses the Doom palette for the color table lookup.
With this option another palette can be used, \fIpalette\fR can
be \fBdoom\fR, \fBheretic\fR, \fBhexen\fR or \fBstrife\fR.
.TP
.BI \-i \ iwadfile
Use \fIiwadfile\fP as the iwad. By default, wadtex does not load any iwad.
.TP
.B \-m
Don't extract a texture if some of its patches are missing.
By default, a texture is output if and only if at least one of its patches
is found. \fB\-m\fP and \fB\-M\fP are mutually exclusive.
.TP
.B \-M
Extract a texture even if all of its patches are missing.
By default, a texture is output if and only if at least one of its patches
is found. \fB\-m\fP and \fB\-M\fP are mutually exclusive.
.TP
.B \-n
List the composition of the textures to standard output but do not create any
PPM files. Note that this option has no effect on the creation of the .wgc
files. Thus, it's possible to get just the .wgc files and not the PPM files by
specifying both \fB\-n\fP and \fB\-t\fP.
.TP
.B \-p
Preserve case. Prevent the name of the graphic and text files to be forced to
lower case. For example, a texture named \fBFRED\fP would be saved as
.B ./textures/FRED.ppm
and
.B ./textures/FRED.wgc
.TP
.B \-t
Also write the texture composition into a text .wgc file for use
with \fBwadgc\fR(6).
.TP
.BI \-w \ expr
Extract only the textures whose names match the wildcard expression \fIexpr\fP.
By default, all textures are extracted. The matching is done by
calling \fBfnmatch\fP(3) with the \fIflags\fP argument set to 0. The strings
are uppercased before they're passed to fnmatch() so that the matching is
case-insensitive. All usual shell wildcards apply:

.ta 2 12
.nf
	\fB*\fP	matches zero or more characters
	\fB?\fP	matches one character
	\fB[\fIrange\fB]\fR	matches one character in \fIrange\fP
	\fB[!\fIrange\fB]\fR	matches one character not in \fIrange\fP
	\fB\\*\fP	matches \fB*\fP
	\fB\\?\fP	matches \fB?\fP
	\fB\\[\fP	matches \fB[\fP
	\fB\\\\\fP	matches \fB\\\fP
	\fIother\fP	matches itself
.fi

See \fBfnmatch\fP(3) for more details.

.SH ENVIRONMENT
This utility supports DOOMWADDIR and DOOMWAD. See the ENVIRONMENT section of
the xwadtools man page for details.

.SH SUPPORTED GAMES
Doom, Ultimate Doom, Doom][, Final Doom, Heretic, Hexen, Strife.

.SH LIMITS
The number of wads on the command line is limited by the maximum number of open
files per process. Big deal.

The size of textures is limited by the available memory and the magnitude of
the size_t type. On 16-bit systems, textures larger than 64 kB will be skipped.

.SH EXAMPLES
To extract all the textures from gothic2.wad:

.nf
	wadtex \-i doom2.wad gothic2.wad
.fi

To extract \fBBROWN1\fP and friends from doom.wad:

.nf
	wadtex \-i doom.wad -w 'brown*'
.fi

.SH BUGS/ERRORS
If a lump is defined more than once in the same wad (which shouldn't happen
anyway), wadtex uses the first occurrence. I think Doom would use the last
occurrence.

Directory searches are slower than they should be.

.SH AUTHOR
.B wadtex
was written by Udo Munk (munkudo@aol.com).
.LP
Various bug fixes and options \fB\-mMnpw\fP added by Andre Majorel
<amajorel@teaser.fr>.

.SH EXIT VALUE
.ta 5
\fB0\fP	OK.
.br
\fB>0\fP	Could not extract all textures.

.SH SEE ALSO
The Unofficial Doom Specs by Matthew S Fell.
.br
.BR tkwadcad "(6), "
.BR wadgc "(6), "
.BR ppm "(5), "
.BR xwadtools (6)


ChangeLog
See "ChangeLog.old" for Udo's changes.

20120227 bkw:

- wadgc: delete temp files after closing them. They're still being
  created in the wrong place (current dir).

- dmpsmu:
  no longer prints high-ASCII garbage in the Dump command.
  now exits if EOF on stdin, like a proper UNIX program (among
  other things, this makes it more easily scriptable)

- testsuite, made a pretty good start on.

- Add - option to trigcalc (normal I/O rather than termios). Allow
  it to compile for mingw (acts like - always present).

- various minor man page fixes

20120223 bkw:

- Fixed FreeBSD 9.0 compile.

- Added crosswin target. Borrowed a fnmatch() implementation from
  OpenBSD, since mingw-w64 seems to lack one.

20120222 bkw:

- Rename Andre's strlcat() to str_lower_cat(). Some OSes (FreeBSD and
  the MinGW cross compiler I have) define a strlcat() that does
  something different.

  - Only include internal strcasecmp() and strncasecmp() if
  !defined(HAVE_STRCASECMP). I don't have an exhaustive knowledge of
  which platforms do and don't provide this function, so the Makefile
  will define HAVE_STRCASECMP by default, with a note to the user.

- lswad's -F now correctly accepts FF_START as start-of-flats marker. Also
  added a warning if -F is used, and there are flats, but no F_END.

- Upgraded deutex to Amajorel Andre's 4.4.902, from here:
  http://www.teaser.fr/~amajorel/deutex/fungus/deutex-4.4.902.tar.gz
  I've made only one minor change to his code: the default log file
  is now /dev/null (can still set it with -log option, if you really
  want).

- Added John Carmack's old spitwad tool, from dmutils.zip. Find the
  original version here:
  http://www.gamers.org/pub/idgames/historic/dmutils.zip

20120221 bkw:

- added a generic man page "doom.6". Too many source ports don't
  supply a man page. Some of them don't supply any documentation
  on command line options at all (or it'll only be documented on
  their web site).

- man pages now installed gzipped by default. "make install GZIP_MAN=no"
  disables.

- binaries are no longer linked with -s (stripped). However, they're
  installed that way by default. "make install STRIP_BINS=no" to
  install unstripped binaries.

20120220 bkw:

- wadcat, mkqmap, mkpopal, mkgipal, mkxppal, mktran were all
  segfaulting if their options that require an argument were given at
  the end of the command line (e.g. "mkqmap -c"), added NULL checks

20120219 bkw:

- New utilities wadwhich, wadconvgame, wadconvgfx

- "make install" now installs the docs (to DOCDIR)

- "make install" now installs mergewad and tkwadcad

20120215 bkw:

- All of Andre Amajorel's patches have been integrated.

- All the tools now compile and run correctly on x86_64 Linux (so far
  as I've been able to determine). This took some effort!

- Makefile.lin has been renamed to Makefile. The other Makefile.* remain,
  but they haven't been updated for this release. I will accept patches
  from anyone who wants to use this on non-Linux systems, but I don't
  have the resources to do any testing on anything but Linux.

- Makefile now supports PREFIX and DESTDIR, should make life easier for
  packagers.

- XForms is now optional, can build with "make USE_XFORMS=no". It's only
  used for xew, so xew won't be built if XForms is disabled.

- The acc utility is now installed as acc-raven, to avoid conflicts with
  the acc utilities provided by VaVoom and ZDoom.

- libfindwad: new library that searches for WAD files in a uniform manner.
  Tools are slowly being updated to use it. See ENVIRONMENT section of
  xwadtools manpage for usage details; see include/findwad.h for API
  and lib/findwad.c for implementation.
  Currently these tools support libfindwad:
  dm2au  wadps  wadwhat  lswad  wadsprit  wadpat  wadtex  wadext
  wadtxls  wadflat

- dm2au:
  New options -r (raw audio output) and -w (wav output).
  Output file can be specified, and can be - for standard output.
  Man page contains a few examples for using dm2au with sox.

- New utility midi3mus. Converts MIDI files to MUS lumps. Written by natt
  on the DoomWorld forums, thread:
  http://www.doomworld.com/vb/doom-editing/55814-midi2mus-64-bit-compatible/

- slige now falls back to $LIBDIR/slige/slige.cfg if no -config or
  $SLIGECONFIG is given and if slige.cfg doesn't exist in the current
  directory.

- wadps: Default game changed to doom2. Behaviour for
  doom II wads now matches doom I, when no level is selected (all maps
  are included in the output; previously the output was just empty).

- lswad: Added --help/-? option. Needed because its default action is now
  to list the contents of the default IWAD.

- wadtxls: Output now sent to stdout if no -o is given.
  Episode and Mission (-e and -m) arguments are no longer required; they
  default to E1M1 for doom 1 and MAP01 for doom 2.

- raw2sfx: outfile argument can be - for stdout. Warn if there's a RIFF or id
  sfx header, or if the input's really short.

- various compiler warnings have been squelched. Quite a few are left.



trunk/wadtex/wadtex.c
/************************************************************************/
/*      Copyright (C) 1998, 1999 by Udo Munk (munkudo@aol.com)          */
/*      Copyright (C) 1998, 2000 by Andre Majorel (amajorel@teaser.fr)  */
/*                                                                      */
/*      Permission to use, copy, modify, and distribute this software   */
/*      and its documentation for any purpose and without fee is        */
/*      hereby granted, provided that the above copyright notice        */
/*      appears in all copies and that both that copyright notice and   */
/*      this permission notice appear in supporting documentation.      */
/*      The author and contibutors make no representations about the    */
/*      suitability of this software for any purpose. It is provided    */
/*      "as is" without express or implied warranty.                    */
/************************************************************************/

/*
 *      Tool to extract wall textures from a WAD file into ppm files
 */

/*
 * 1.3 (AYM 1998-12-22)
 * - Fixed bug that made the first row and the first column appear pink.
 * - Fixed bug in the handling of patches with negative Y offsets.
 * - The names of the output files are now forced to lower case
 *   (rationale: since case is not significant in wad directories, why
 *   not use the most convenient case)
 * - Added the "-p" option to inhibit the above feature.
 *
 * 1.4 (AYM 1999-01-14)
 * - A missing patch is not a fatal error anymore. wadtexext can now be
 *   used on texture pwads.
 * - Added options "-m" and "-M" to tweak the behaviour in face of missing
 *   patches.
 *
 * 1.5 (UM 1999-02-15)
 * - Added support for Heretic and Hexen.
 *
 * 1.6 (AYM 1999-03-07)
 * - Fixed segfault with large textures. Textures that are too large to
 *   fit in the buffer are now cleanly clipped with a warning.
 * - Increased the size of the buffer from 256x200 to 1024x256. I
 *   suspect that 256 might turn out to be insufficient, though... I
 *   must have a bad case of the C programmer's disease... And I should
 *   clip by (width x height) instead of dumbly clipping by width and
 *   height separately.
 *     (UM 1999-03-25)
 * - Added support for Strife.
 *
 * 1.7 (UM 1999-08-23)
 * - Added option -i iwad, so that patches missing in a PWAD can be
 *   used from the IWAD, to complete a texture.
 *
 * 1.8 (AYM 2000-10-26)
 * - Added option "-n" to just list the composition of textures without
 *   saving them to file.
 * - Now accepts any number (including zero) of pwads.
 * - Changed the mode of textures/ from 0755 to 0777 on the theory that
 *   limiting permissions is the job of umask, not of the application.
 * - Fixed most of the counterintuitive quirks w.r.t. to the handling of
 *   wads, notably the fact that the pwad had to have a PNAMES and
 *   TEXTURE1. It's now possible to extract textures from a pwad that
 *   replaces just patches like dmspac.zip.
 * - Program name is now hard-coded because I felt it didn't make much
 *   sense to have wadtex print "stupid version 1.8" on startup just
 *   because someone renamed the binary to "stupid".
 * - Removed the arbitrary limit of 1024x256 on image size by making the
 *   texture buffer dynamically allocated. It's now possible to extract
 *   all textures from the shareware Hacx wad, including "AD" that is
 *   2048 wide. I just knew I had the C programmer's disease.
 * - Error messages now printed through new err() function so that they
 *   don't cut stdout messages in the middle when redirected.
 * - Added option "-w" to extract only some of the textures.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <sys/stat.h>

#ifdef HAVE_FNMATCH
#include <fnmatch.h>
#else
#include "obsd_fnmatch.h"
#endif

#include "sysdep.h"
#include "strfunc.h"
#include "wad.h"
#include "lump_dir.h"
#include "wadfile.h"
#include "findwad.h"

#define VERSION "1.8"

/* item in a linked list of wads */
typedef struct listitem_s {
	struct listitem_s *next;
	char *name;
	wadfile_t *wf;
} listitem_t;

const char program[] = "wadtex";

extern unsigned char doom_rgb[];
extern unsigned char heretic_rgb[];
extern unsigned char hexen_rgb[];
extern unsigned char strife_rgb[];
static unsigned char *palette = doom_rgb;

static unsigned char *img_buf;
static size_t        img_buf_sz;

static int preserve_case = 0;
static int tflag;
static int extract = 1;
static char *wildcard;

static enum {
	OIT_IF_ALL,	/* Output the texture iff no patch is missing */
	OIT_IF_ONE,	/* Output the texture iff at least one patch found */
	OIT_ALWAYS	/* Always output the texture */
} mpatch_policy = OIT_IF_ONE;

void err(const char *fmt, ...)
{
	va_list list;

	fflush(stdout);
	fprintf(stderr, "%s: ", program);
	va_start(list, fmt);
	vfprintf(stderr, fmt, list);
	va_end(list);
	fputc('\n', stderr);
}

void usage(const char *option)
{
	if (option)
		err("Unkown option: %s", option);

	printf("Usage: %s [-c pal] [-i iwad] [-m|-M] [-npt] [-w expr] [wadfile ...]\n",
	    program);
	printf("\t-c pal: use color palette pal, where pal can be\n");
	printf("\t        doom, heretic, hexen or strife\n");
	printf("\t-i iwad: use iwad to find patches not found in wadfile\n");
	printf("\t-m: don't output texture if patches are missing\n");
	printf("\t-M: output texture even if all patches are missing\n");
	printf("\t-n: list the composition without creating any PPM files\n");
	printf("\t-p: preserve case of generated files\n");
	printf("\t-t: also write text description of textures for wadgc\n");
	printf("\t-w expr: only extract textures matching wildcard expr\n");
	exit(1);
}

int get_patch(wadfile_t *wf, FILE *fp, void *pnames, short width, short height,
	       short xoff, short yoff, short pnum, const char **patchname)
{
	int		i;
	int		x, y;
	int		n;
	int		*p1;
	char		*p2;
	int		p_num;
	char		name[9];
	unsigned char	*picture;
	short		pwidth;
	int		*columns;
	unsigned char	*post;

	/* initialization */
	memset(&name[0], 0, 9);

	/* number of patches in PNAMES */
	p1 = (int *) pnames;
	p_num = *p1++;
	swaplong(&p_num);
	if (pnum > p_num) {
	  err("texture uses patch %d, but PNAMES has %d patches only", pnum, p_num);
	  exit(1);
	}

	/* lookup name for that patch */
	p2 = ((char *) p1) + 8 * pnum;
	*patchname = (const char *) p2;
	strncpy(&name[0], p2, 8);
	for (i = 0; i < 8; i++) {
		name[i] = toupper(name[i]);
	}

	/* and get that picture lump */
	if ((picture=(unsigned char *)get_lump_by_name(wf, &name[0])) == NULL) {
		return 1;
	}

	/* copy picture data into pixel buffer */
	pwidth = *((short *) picture);
	swapint(&pwidth);
	printf("    Got patch %s, width %d, x=%d y=%d\n", &name[0], pwidth,
			xoff, yoff);
	columns = (int *) (picture + 8);
	for (x = 0; x < pwidth; x++) {
	  swaplong(columns);
	  post = picture + *columns++;
	  y = (int) *post++;
	  while (y != 255) {
	    n = (int) *post++;
	    post++;	/* first byte unused */
	    for (i = 0; i < n; i++) {
		if ((y+yoff+i >= 0) && (x+xoff >= 0) &&
		    (y+yoff+i < height) && (x+xoff < width))
		  img_buf[(y+yoff+i) * width + (x+xoff)] = *post;
		post++;
	    }
	    post++;	/* last byte unused */
	    y = (int) *post++;
	  }
	}

	/* text file too? */
	if (tflag) {
	  fprintf(fp, "    %s %d %d\n", &name[0], xoff, yoff);
	}

	/* free picture lump memory */
	free(picture);
	return 0;
}

void write_ppm(const char *name, short width, short height)
{
	FILE		*fp;
	char		fn[50];
	int		i, j;
	unsigned char	pixel;

	strcpy(&fn[0], "textures/");
	if (preserve_case)
	  strcat(&fn[0], name);
	else
	  str_lower_cat(&fn[0], name);
	strcat(&fn[0], ".ppm");
	if ((fp = fopen(&fn[0], "wb")) == NULL) {
		err("can't open %s for writing", &fn[0]);
		exit(1);
	}

	fprintf(fp, "P6\n");
	fprintf(fp, "# CREATOR: %s Release %s\n", program, VERSION);
	fprintf(fp, "%hd %hd\n", width, height);
	fprintf(fp, "255\n");

	for (j = 0; j < height; j++) {
		for (i = 0; i < width; i++) {
			pixel = img_buf[j * width + i];
			fputc((int) palette[pixel * 3],     fp);
			fputc((int) palette[pixel * 3 + 1], fp);
			fputc((int) palette[pixel * 3 + 2], fp);
		}
	}

	if (fclose(fp)) {
		err("%s: %s", fn, strerror(errno));
		exit(1);
	};
}

/*
 * returns 0 on success, non-zero if not all textures were properly extracted.
 */
int decompile(listitem_t *wadlist, char *section, void *texture, void *pnames)
{
	int	rc = 0;
	int	i, j;
	int	patches_found;
	int	*p1;
	char	*p2;
	short	*p3;
	int	num_tex;
	int	off_tex;
	short	dummy;
	short	width, height, num_pat;
	short	xoff, yoff, pnum;
	char	tname[9];
	FILE	*fp = NULL;
	char	fn[50];

	/* initialize */
	memset(&tname[0], 0, sizeof tname);

	/* get number of textures in texture lump */
	p1 = (int *) texture;
	num_tex = *p1;
	swaplong(&num_tex);
	printf("  found %d textures\n", num_tex);

	/* loop over all textures */
	for (i = 0; i < num_tex; i++) {
 
		/* offset for texture in lump */
		p1 = (int *) texture + i + 1;
		off_tex = *p1;
		swaplong(&off_tex);
		p2 = (char *) texture + off_tex;

		/* get name of texture */
		strncpy(&tname[0], p2, 8);
		{
		  int n;
		  for (n = 0; n < sizeof tname; n++)
			  tname[n] = toupper(((unsigned char *) tname)[n]);
		}
		if (wildcard != NULL && fnmatch(wildcard, tname, 0) != 0)
			continue;

		/* get texture composition */
		p3 = (short *) (p2 + 8);
		dummy = *p3++;		/* always 0 */
		dummy = *p3++;		/* always 0 */
		width = *p3++;
		swapint(&width);
		height = *p3++;
		swapint(&height);
		dummy = *p3++;		/* always 0 */
		dummy = *p3++;		/* always 0 */
		num_pat = *p3++;
		swapint(&num_pat);

		printf("  decompiling texture %s, %d patches, %dx%d\n",
		  &tname[0], num_pat, width, height);

		/* grow the texture buffer if necessary */
		{
		  unsigned long newsize = width * height;
		  if ((size_t) newsize != newsize) {
		    err("Texture too big, size_t overflow (%hd x %hd). Skipping.",
			width, height);
		    rc = 1;
		    continue;
		  }
		  if (newsize > img_buf_sz) {
		    unsigned char *newbuf = realloc(img_buf, newsize);
		    if (newbuf == NULL) {
		      err("Texture too big, not enough memory (%d x %d). Skipping.",
			  width, height);
		      rc = 1;
		      continue;
		    }
		    img_buf_sz = newsize;
		    img_buf    = newbuf;
		  }
		}

		/* want text file too? */
		if (tflag) {
		  strcpy(&fn[0], "textures/");
		  if (preserve_case)
		    strcat(&fn[0], &tname[0]);
		  else
		    str_lower_cat(&fn[0], &tname[0]);
		  strcat(&fn[0], ".wgc");
		  if ((fp = fopen(&fn[0], "w")) == NULL) {
		    err("can't open %s for writing", &fn[0]);
		    exit(1);
		  }
		  fprintf(fp, "%s_START\n", section);
		  fprintf(fp, "  %s %d %d %d\n", &tname[0], width, height,
		      num_pat);
		}

		/* clear pixel buffer */
		memset(img_buf, 255, img_buf_sz);

		/* loop over all patches for the texture */
		patches_found = 0;
		for (j = 0; j < num_pat; j++) {
		  const char *patchname = NULL;
		  listitem_t *wad;

		  xoff = *p3++;
		  swapint(&xoff);
		  yoff = *p3++;
		  swapint(&yoff);
		  if (yoff < 0)
                    yoff = 0;
		  pnum = *p3++;
		  swapint(&pnum);
		  dummy = *p3++;	/* always 1, stepdir? what is that?? */
		  dummy = *p3++;	/* always 0 for colormap 0 */
		  for (wad = wadlist; wad != NULL; wad = wad->next) {
		    if (get_patch(wad->wf, fp, pnames, width, height,
			  xoff, yoff, pnum, &patchname) == 0) {
		      patches_found++;
		      break;
		    }
		  }
		  if (wad == NULL)
		    printf("    Warning: patch \"%.8s\" is missing\n",
			patchname);
		}

		if (patches_found == 0 && mpatch_policy != OIT_ALWAYS) {
		  printf("    Warning: skipping void texture \"%s\".\n", tname);
		  rc = 1;
		}
		else if (patches_found < num_pat && mpatch_policy == OIT_IF_ALL) {
		  printf("    Warning: skipping incomplete texture \"%s\".\n", tname);
		  rc = 1;
		}
		else
		  if (extract)
		    write_ppm(&tname[0], width, height);

		/* terminate and close the text file too */
		if (tflag) {
		  fprintf(fp, "%s_END\n", section);
		  if (fclose(fp)) {
		    err("%s: %s", fn, strerror(errno));
		    exit(1);
		  }
		}
	}
	return rc;
}

int main(int argc, char **argv)
{
	int        rc = 0;		/* Exit code */
	char       *s;
	listitem_t *wadlist;		/* Linked list of open wads */
	char       *iwad = NULL;	/* Filename of the iwad (or NULL) */
	void       *pnames;		/* The contents of the PNAMES lump */
	void       *texture1;		/* The contents of the TEXTURE1 lump */
	void       *texture2;		/* The contents of the TEXTURE2 lump */

	printf("%s version %s\n\n", program, VERSION);

	/* process options */
	while ((--argc > 0) && (**++argv == '-')) {
	  for (s = *argv+1; *s != '\0'; s++) {
		switch (*s) {
		case 'c':
			argc--;
			argv++;
			if (!strcmp(*argv, "doom"))
				palette = doom_rgb;
			else if (!strcmp(*argv, "heretic"))
				palette = heretic_rgb;
			else if (!strcmp(*argv, "hexen"))
				palette = hexen_rgb;
			else if (!strcmp(*argv, "strife"))
				palette = strife_rgb;
			else
				usage(NULL);
			break;

		case 'i':
			argc--;
			argv++;
			iwad = *argv;
			break;

		case 'm':
			mpatch_policy = OIT_IF_ALL;
			break;

		case 'M':
			mpatch_policy = OIT_ALWAYS;
			break;

		case 'n':
			extract = 0;
			break;

		case 'p':
			preserve_case++;
			break;

		case 't':
			tflag++;
			break;

		case 'w':
			argc--;
			argv++;
			wildcard = *argv;
			{
			  char *p;
			  for (p = wildcard; *p != '\0'; p++)
				  *p = toupper(*((unsigned char *) p));
			}
			break;

		default:
			usage(--s);
		}
	  }
	}

	if (argc == 0 && iwad == NULL) {
	  err("No wad given");
	  exit(1);
	}

	/* build list of wads, last pwad first, iwad last */
	wadlist = NULL;
	if (iwad != NULL) {
	  wadlist = malloc(sizeof *wadlist);
	  if (wadlist == NULL) {
	    err("Not enough memory");
	    exit(1);
	  }
	  wadlist->name = iwad;
	  wadlist->wf   = open_wad(fw_find_wad(iwad));
	  wadlist->next = NULL;
	}
	for (; argc > 0; argv++, argc--) {
	  listitem_t *newitem = malloc(sizeof *newitem);
	  if (newitem == NULL) {
	    err("Not enough memory");
	    exit(1);
	  }
	  /* insert at the beginning of the list */
	  newitem->name = *argv;
	  newitem->wf   = open_wad(fw_find_wad(*argv));
	  newitem->next = wadlist;
	  wadlist = newitem;
	}

	/* locate PNAMES, TEXTURE1 and TEXTURE2 */
	pnames   = NULL;
	texture1 = NULL;
	texture2 = NULL;
	{
	  listitem_t *wad;

	  for (wad = wadlist; wad != NULL; wad = wad->next) {
	    if (pnames == NULL)
	      pnames = get_lump_by_name(wad->wf, "PNAMES");
	    if (texture1 == NULL)
	      texture1 = get_lump_by_name(wad->wf, "TEXTURE1");
	    if (texture2 == NULL)
	      texture2 = get_lump_by_name(wad->wf, "TEXTURE2");
	  }
	}
	if (pnames == NULL) {
		err("No PNAMES lump found");
		exit(1);
	}
	if (texture1 == NULL) {
		err("No TEXTURE1 lump found");
		exit(1);
	}
	    
	/*
	 * make textures directory, ignore errors, we'll handle that later
	 * when we try to write the graphics files into it
	 */
	if (extract || tflag)
#ifdef __MINGW32__
		mkdir("textures");
#else
		mkdir("textures", 0777);
#endif

	/* decompile the textures */
	printf("Got TEXTURE1 lump, decompiling...\n");
	{
	  int r = decompile(wadlist, "TEXTURE1", texture1, pnames);
	  if (r > rc)
		  rc = r;
	  if (texture2 != NULL) {
		  printf("Got TEXTURE2 lump, decompiling...\n");
		  r = decompile(wadlist, "TEXTURE2", texture2, pnames);
		  if (r > rc)
			  rc = r;
	  }
	}

	/* clean up and done */
	if (pnames != NULL)
	  free(pnames);
	if (texture1 != NULL)
	  free(texture1);
	if (texture2 != NULL)
	  free(texture2);
	{
	  listitem_t *wad, *next;

	  for (wad = wadlist; wad != NULL;) {
	    next = wad->next;
	    if (wad->wf != NULL)
	      close_wad(wad->wf);
	    free(wad);
	    wad = next;
	  }
	}
	if (img_buf != NULL)
	  free(img_buf);
	return rc;
}
